proposals/2024-01-29_native_histograms_text_format.md

```diff
@@ -0,0 +1,110 @@
+## Native Histograms Text Format
+
+* **Owners:**
+  * Chris Marchbanks [@csmarchbanks](https://github.com/csmarchbanks)
+
+* **Implementation Status:** Not implemented
+
+* **Related Issues and PRs:**
+  * [Native Histogram Support in client_python](https://github.com/prometheus/client_python/issues/918)
+  * [OpenMetrics Protobuf format PR](https://github.com/OpenObservability/OpenMetrics/pull/256)
+
+* **Other docs or links:**
+  * [Design Doc for choosing a proposal](https://docs.google.com/document/d/1qoHf24cKMpa1QHskIjgzyf3oFhIPvacyJj8Tbe6fIrY/edit#heading=h.5sybau7waq2q)
+
+> TL;DR: This design doc is proposing a format for exposing native histograms in the OpenMetrics text format.
+
+## Why
+
+Today it is only possible to export native histograms using the classic (not OpenMetrics) Protocol Buffers (protobuf) scrape format. Many users prefer the text format, and some client libraries, such as the Python client, want to avoid adding a dependency on protobuf. Prometheus content negotiation prefers OpenMetrics to the classic Prometheus text based format, therefore to support native histograms in Prometheus the OpenMetrics text format will also need to support native histograms.
+
+There is already an open pull request (see Related Issues and PRs above) to add support for native histograms to OpenMetrics, and during a [dev summit in 2022](https://docs.google.com/document/d/11LC3wJcVk00l8w5P3oLQ-m3Y37iom6INAMEu2ZAGIIE/edit#bookmark=id.c3e7ur6rn5d2) there was consensus we would continue to support the text format for new features as well. Including native histograms as part of the text format shows commitment to that consensus.
+
+### Pitfalls of the current solution
+
+Prometheus client libraries such as Python do not want to require a dependency on protobuf in order to expose native histograms, and in some languages protobuf is painful to use. Gating native histograms only to clients/users willing to use protobuf hurts adoption of native histograms, therefore, we would like a way to represent a native histogram in the text based format.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Support native histograms in the text format
```

Comment:
Is support of custom (explicit) bucketing a goal of this proposal? (I hope so! 🙈 )

Reply:
Correct me if I have missed something, but I think the current proposal for that is starting with classic histograms and storing them as a custom native histogram internally? If so this proposal supports that. This proposal does not support defining custom buckets outside of the classic histogram format, but could be done fairly trivially, and alongside any protobuf changes, in OpenMetrics 1.2.

Reply:
Yes, current proposal for custom buckets is to upconvert on scrape. So it's super useful for legacy apps where you don't have the code/compiler anymore.

Reply:
Indeed. I wouldn't see custom buckets directly in the text format a strict necessity. On the contrary, the classic representation is way more readable, and it will result in the same NHCB upon ingestion.

Reply:
I don't think I fully agree, there are benefits to nhcb on the text exposition format:

* nhcb would make SDKs overhead smaller (e.g. see [Otel Java benchmarks](https://opentelemetry.io/blog/2024/java-metric-systems-compared/) on histograms - while the whole benchmark could be argued on, the classic histograms have some overhead as they are treated as separate lines on scrape)
* nhcb would make server scrape cheaper (probably not significant though?)

But maybe you argue proto should be default for performance? But it was just said that some languages would never do proto..

Anyway, I am happy to scope this down here, do you mind @csmarchbanks to add nhcb as explicit No-Goal item? Thanks! 🤗

Reply:
In the old days, we usually said "text format is for simple cases, if you want industry strength performance, use protobuf" (which ironically didn't really work out well back then because the Go protobuf implementation was so wasteful).

But I think this argument still makes some sense. You could say that languages that just don't want protobuf at all are probably also the languages you wouldn't use if performance is a core concern (Python being a good example for this case).

Reply:
Added as a non-goal as requested. Specifically:
```
Support for custom buckets as part of the native histogram structure. If needed by a consumer, custom buckets will need to be parsed from classic buckets.
```
to make it clear how consumers can use custom buckets if desired.